core/services/gas_updater.go

```diff
@@ -57,91 +70,292 @@ type gasUpdater struct {
 	gu := &gasUpdater{
+		StartStopOnce:           utils.StartStopOnce{},
 		store:                   store,
-		rollingBlockHistory:     make([]*types.Block, 0),
-		rollingBlockHistorySize: int(store.Config.GasUpdaterBlockHistorySize()),
+		rollingBlockHistory:     make([]Block, 0),
+		rollingBlockHistorySize: int64(store.Config.GasUpdaterBlockHistorySize()),
 		blockDelay:              int64(store.Config.GasUpdaterBlockDelay()),
 		percentile:              int(store.Config.GasUpdaterTransactionPercentile()),
+		mb:                      utils.NewMailbox(1),
+		wg:                      new(sync.WaitGroup),
+		ctx:                     ctx,
+		ctxCancel:               cancel,
+		logger:                  logger.CreateLogger(logger.Default.With("id", "gas_updater")),
 	}
+
 	return gu
 }
 
 func (gu *gasUpdater) Connect(bn *models.Head) error {
-	if gu.store.Config.GasUpdaterEnabled() {
-		logger.Debugw("GasUpdater: dynamic gas updates are enabled", "ethGasPriceDefault", gu.store.Config.EthGasPriceDefault())
-	} else {
-		logger.Debugw("GasUpdater: dynamic gas updating is disabled", "ethGasPriceDefault", gu.store.Config.EthGasPriceDefault())
-	}
 	return nil
 }
 
 func (gu *gasUpdater) Disconnect() {
 }
 
-// OnNewLongestChain recalculates and sets global gas price on every head
+// OnNewLongestChain recalculates and sets global gas price if a new head comes
+// in and we are not currently fetching
 func (gu *gasUpdater) OnNewLongestChain(ctx context.Context, head models.Head) {
 	// Bail out as early as possible if the gas updater is disabled so we avoid
 	// any potential undesired side effects. Note that in a future iteration
 	// the GasUpdaterEnabled setting could be modifiable at runtime
-	if !gu.store.Config.GasUpdaterEnabled() {
-		return
+	if gu.State() != utils.StartStopOnce_Started {
+		panic("OnNewLongestChain called on unstarted gas updater")
+	}
+	gu.mb.Deliver(head)
+}
+
+func (gu *gasUpdater) Start() error {
+	if !gu.OkayToStart() {
+		return errors.New("GasUpdater has already been started")
+	}
+	gu.logger.Debugw("GasUpdater: starting")
+	if gu.rollingBlockHistorySize > int64(gu.store.Config.EthFinalityDepth()) {
+		gu.logger.Warnf("GasUpdater: GAS_UPDATER_BLOCK_HISTORY_SIZE=%v is greater than ETH_FINALITY_DEPTH=%v, blocks deeper than finality depth will be refetched on every gas updater cycle, causing unnecessary load on the eth node. Consider decreasing GAS_UPDATER_BLOCK_HISTORY_SIZE or increasing ETH_FINALITY_DEPTH", gu.rollingBlockHistorySize, gu.store.Config.EthFinalityDepth())
+	}
+	ctx, cancel := context.WithTimeout(gu.ctx, maxStartTime)
+	defer cancel()
+	latestHead, err := gu.store.EthClient.HeaderByNumber(ctx, nil)
+	if err != nil {
+		logger.Warnw("GasUpdater: initial check for latest head failed", "err", err)
+	} else {
+		gu.logger.Debugw("GasUpdater: got latest head", "number", latestHead.Number, "blockHash", latestHead.Hash.Hex())
+		gu.FetchBlocksAndRecalculate(ctx, *latestHead)
+	}
+	gu.wg.Add(1)
```

Comment:
I think a done channel would be less error prone here since we only have one goroutine

Reply:
The context allows us to cancel requests when we close it so we avoid waiting.

Reply:
Which part is error-prone?

Reply:
Its more of a nit, but with wait groups you have to make sure the wg.Done and wg.Adds match up and you don't have to worry about that with a single channel